File: mobile/src/services/api/sessions.js
import { apiRequest } from './client';

function toSessionShape(payload) {
  return {
    id: payload.id,
    exerciseId: payload.exerciseId ?? payload.exercise_id,
    difficulty: payload.difficulty,
    status: payload.status,
    targetSets: payload.targetSets ?? payload.target_sets ?? 0,
    targetReps: payload.targetReps ?? payload.target_reps ?? 0,
    externalLoadKg: payload.externalLoadKg ?? payload.external_load_kg ?? null,
    bodyWeightKg: payload.bodyWeightKg ?? payload.body_weight_kg ?? null,
    totalReps: payload.totalReps ?? payload.total_reps ?? 0,
    avgFormScore: payload.avgFormScore ?? payload.avg_form_score ?? null,
    durationSeconds: payload.durationSeconds ?? payload.duration_seconds ?? null,
    createdAt: payload.createdAt ?? payload.created_at,
    startedAt: payload.startedAt ?? payload.started_at ?? null,
    endedAt: payload.endedAt ?? payload.ended_at ?? null,
  };
}

function toFeedbackShape(payload) {
  return {
    id: payload.id,
    sessionId: payload.sessionId ?? payload.session_id,
    exerciseId: payload.exerciseId ?? payload.exercise_id,
    overallRating: payload.overallRating ?? payload.overall_rating ?? 0,
    summaryText: payload.summaryText ?? payload.summary_text ?? '',
    errorsCount: payload.errorsCount ?? payload.errors_count ?? 0,
    topErrors: payload.topErrors ?? payload.top_errors ?? [],
    errorBreakdown: payload.errorBreakdown ?? payload.error_breakdown ?? {},
    actionItems: payload.actionItems ?? payload.action_items ?? [],
    confidenceOverall: payload.confidenceOverall ?? payload.confidence_overall ?? null,
    signalsUsed: payload.signalsUsed ?? payload.signals_used ?? {},
    version: payload.version,
    generatedAt: payload.generatedAt ?? payload.generated_at,
    updatedAt: payload.updatedAt ?? payload.updated_at,
  };
}

export async function createSession({
  exerciseId,
  targetSets = 1,
  targetReps = 1,
  difficulty = 'beginner',
  externalLoadKg = null,
  bodyWeightKg = null,
}) {
  const payload = await apiRequest('/sessions', {
    method: 'POST',
    auth: true,
    body: { exerciseId, targetSets, targetReps, difficulty, externalLoadKg, bodyWeightKg },
  });
  return toSessionShape(payload);
}

export async function getSession(sessionId) {
  const payload = await apiRequest(`/sessions/${sessionId}`, { auth: true });
  return toSessionShape(payload);
}

export async function listSessions() {
  const payload = await apiRequest('/sessions', { auth: true });
  return Array.isArray(payload) ? payload.map(toSessionShape) : [];
}

export async function startSession(sessionId) {
  const payload = await apiRequest(`/sessions/${sessionId}/start`, {
    method: 'POST',
    auth: true,
  });
  return toSessionShape(payload);
}

export async function endSession(sessionId, summary = {}) {
  const payload = await apiRequest(`/sessions/${sessionId}/end`, {
    method: 'POST',
    auth: true,
    body: {
      totalReps: summary.totalReps ?? summary.total_reps ?? 0,
      avgFormScore: summary.avgFormScore ?? summary.avg_form_score ?? null,
      externalLoadKg: summary.externalLoadKg ?? summary.external_load_kg ?? null,
      bodyWeightKg: summary.bodyWeightKg ?? summary.body_weight_kg ?? null,
    },
  });
  return toSessionShape(payload);
}

export async function getSessionFeedback(sessionId) {
  const payload = await apiRequest(`/sessions/${sessionId}/feedback`, { auth: true });
  return toFeedbackShape(payload);
}

File: backend/coachvision/api/sessions.py
"""Workout session lifecycle endpoints."""

from datetime import datetime, timezone
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session as DbSession

from .deps import get_current_user
from .schemas import CreateSessionRequest, EndSessionRequest, SessionFeedbackResponse, SessionResponse
from ..db.models import Exercise, Session, SessionFeedback, User
from ..db.session import get_db

router = APIRouter(prefix="/sessions")


@router.post("", response_model=SessionResponse)
def create_session(
    payload: CreateSessionRequest,
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    exercise = db.get(Exercise, payload.exercise_id)
    if not exercise:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Exercise not found")

    session = Session(
        user_id=current_user.id,
        exercise_id=payload.exercise_id,
        difficulty=payload.difficulty,
        status="planned",
        target_sets=payload.target_sets,
        target_reps=payload.target_reps,
    )
    db.add(session)
    db.commit()
    db.refresh(session)
    return SessionResponse.model_validate(session)


@router.get("", response_model=list[SessionResponse])
def list_sessions(
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[SessionResponse]:
    rows = db.scalars(
        select(Session)
        .where(Session.user_id == current_user.id, Session.status == "completed")
        .order_by(Session.ended_at.desc(), Session.created_at.desc())
    ).all()
    return [SessionResponse.model_validate(row) for row in rows]


@router.get("/{session_id}", response_model=SessionResponse)
def get_session(
    session_id: UUID,
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    session = db.get(Session, session_id)
    if not session or session.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Session not found")
    return SessionResponse.model_validate(session)


@router.get("/{session_id}/feedback", response_model=SessionFeedbackResponse)
def get_session_feedback(
    session_id: UUID,
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionFeedbackResponse:
    session = db.get(Session, session_id)
    if not session or session.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Session not found")
    row = db.scalars(
        select(SessionFeedback).where(
            SessionFeedback.session_id == session_id,
            SessionFeedback.user_id == current_user.id,
        )
    ).first()
    if row is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session feedback not found",
        )
    return SessionFeedbackResponse.model_validate(row)


@router.post("/{session_id}/start", response_model=SessionResponse)
def start_session(
    session_id: UUID,
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    session = db.get(Session, session_id)
    if not session or session.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Session not found")
    session.status = "active"
    if not session.started_at:
        session.started_at = datetime.now(timezone.utc)
    session.updated_at = datetime.now(timezone.utc)
    db.add(session)
    db.commit()
    db.refresh(session)
    return SessionResponse.model_validate(session)


@router.post("/{session_id}/end", response_model=SessionResponse)
def end_session(
    session_id: UUID,
    payload: EndSessionRequest | None = None,
    db: DbSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    session = db.get(Session, session_id)
    if not session or session.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Session not found")
    if session.status == "completed":
        from ..services.session_feedback_generator import ensure_session_feedback_for_completed_workout

        ensure_session_feedback_for_completed_workout(db, session)
        db.commit()
        db.refresh(session)
        return SessionResponse.model_validate(session)

    session.status = "completed"
    now_utc = datetime.now(timezone.utc)
    session.ended_at = now_utc
    if session.started_at:
        session.duration_seconds = max(0, int((now_utc - session.started_at).total_seconds()))
    if payload is not None and payload.total_reps is not None:
        session.total_reps = payload.total_reps
    if payload is not None and payload.avg_form_score is not None:
        session.avg_form_score = payload.avg_form_score
    session.updated_at = now_utc
    db.add(session)
    from ..services.session_feedback_generator import ensure_session_feedback_for_completed_workout

    ensure_session_feedback_for_completed_workout(db, session)
    db.commit()
    db.refresh(session)
    from ..services.fatigue_post_session import run_after_completed_session

    run_after_completed_session(db, session.id)
    return SessionResponse.model_validate(session)

